# Build in house allele frequency and genotype and phenotype database for your project

Assume you have  sequenced a lot of DNA samples, and you want to build a local database for annotation and screening. Here is a protocol to tell you how to build in house databases and how to use them with bcfanno.



1. Generate VCFs of these samples.

   Assume you have enough computation resource and hard disk to generate and restore VCFs of your samples. Most of genetic variants callers (like GATK, samtools, varscan) only generate variants with related information. However, we strongly suggest to export genotype for each position, include the REFs, in the VCFs when you do variants calling. To store all the genotypes for each sample will cost lots of extra CPU and hard disk, but it's worth to do that if more accuracy database be expected. And the hard disk problem could be highly avoid by merge two or more VCFs into a multi-samples VCF. 

   Actually some advanced variant caller like GATK HaplotypeCaller could export all the genotype information for each genome position with properly defined parameters. Here is an example.

   [example.png]

   â€‹

2. Merge VCFs into one big file.

   `bcftools merge -m both sample1.vcf.gz sample2.vcf.gz... sampleN.vcf.gz -O b -o merged.bcf`



3. Calculate the allele frequency and remove extra sample information.

```
   allele_freqs_count -h
   A program to calculate the allele frequency in population for each position.
   Usage:
    vcf_allele_freq [options] input.vcf.gz
    Options:
     -ac   allele count tag name, default is AlleleCount
     -af   allele frequency tag name, default is AlleleFreq
     -hwe  tag name for Fisher exact test of Hardy-Weinberg Equilibrium
     -sam  affected sample list for each allele, sample seperated by /
     -gen  sample counts for homozygous ref, heterozygous and homozygous alts, the format is "AA | Aa | aa"
     -O    output format [b|z]
     -o    output file
     -force_skip_uncover  force skip the uncovered positions, default treat as reference

Note:
1. For parameter -gen, the tag format is [AA|Aa|aa], 'A' stand for wild allele, and 'a' stand for mutated
   allele(s). All alternative alleles, if more than one alt allele like mnps, should be count into 'a'.
   Version: 8699629 + htslib 1.4.1
   Homepage: https://github.com/shiquan/small_projects

```